pkgs/build-support/rust/hooks/default.nix

```diff
@@ -27,16 +27,29 @@ let
   rustBuildPlatform = rust.toRustTarget stdenv.buildPlatform;
   rustTargetPlatform = rust.toRustTarget stdenv.hostPlatform;
   rustTargetPlatformSpec = rust.toRustTargetSpec stdenv.hostPlatform;
-in {
-  cargoBuildHook = callPackage ({ }:
+  cargoBuildHookFlexible = { cargoPurityFlag }: callPackage ({ }:
     makeSetupHook {
       name = "cargo-build-hook.sh";
       deps = [ cargo ];
       substitutions = {
-        inherit ccForBuild ccForHost cxxForBuild cxxForHost
+        inherit ccForBuild ccForHost cxxForBuild cxxForHost cargoPurityFlag
           rustBuildPlatform rustTargetPlatform rustTargetPlatformSpec;
       };
     } ./cargo-build-hook.sh) {};
+in {
+
+  # "requires that the Cargo.lock file is up-to-date. If the lock file
+  # is missing, or it needs to be updated, Cargo will exit with an
+  # error. The --frozen flag also prevents Cargo from attempting to
+  # access the network to determine if it is out-of-date."
+  cargoBuildHook = cargoBuildHookFlexible { cargoPurityFlag = "--frozen"; };
+
+  # "Prevents Cargo from accessing the network for any reason.  Beware
+  # that this may result in different dependency resolution than
+  # online mode. Cargo will restrict itself to crates that are
+  # downloaded locally, even if there might be a newer version as
+  # indicated in the local copy of the index."
+  cargoBuildHookOffline = cargoBuildHookFlexible { cargoPurityFlag = "--offline"; };
```

Comment:
Not sure if we want to expose this or not. Could we easily inline this into build-rust-package?

Reply:
Ah, we can.

The problem was that `cargoBuildHook` is already exposed and didn't take an argument, so it has to continue to be usable that way.

For some reason I didn't notice that `cargoBuildHook` is a `callPackage` -- so I can add the argument there and give it a default argument that preserves the current behavior.

See 39b524943abde4080a7429ba9b0f93c2df39c5b9.